### PR TITLE
Add generator for collection resource optionally including basic metadata

### PR DESCRIPTION
**Pull request by elrayle**
### Description

This PR...
* adds a generator for creating a collection resource class
* adds to dassie: CollectionResource, a collection resource class with basic metadata created by the generator
* adds to .internal_test app generation process: generate CollectionResource with basic metadata
* adds a factory to create an instance of CollectionResource
* updates collection_controller_with_resource_spec.rb to use the generated CollectionResource class

### Collection Resource Generator

Generates:
1. Creates a collection model and spec.
2. Creates a collection form and spec.
3. Creates a collection indexer and spec.
4. Creates a collection metadata config.
5. Sets this to be the collection class.

The generated collection resource classes optionally include `basic_metadata`.

### Add CollectionResource in test app

The test generator was updated to create a collection resource class named "CollectionResource".  The config that is generated to set this as 'the' collection class is commented out in `/config/initializers/hyrax.rb` by the test generator.  

For now, this allows the ActiveFedora `::Collection` class to continue to be 'the' collection class.

_NOTE: At this time, Hyrax supports a single class that is 'the' collection class.  It is configured in `config/initializers/hyrax.rb` by setting `config.collection_model =`._

### Potential Future Work

This pattern allows for future work that would update Hyrax to allow multiple collection resource classes each with their own metadata.

### Updates CollectionController test for resources

Updates `spec/controllers/hyrax/dashboard/collection_controller_with_resource_spec.rb` to use CollectionResource class for testing.  This verifies the generated class and allows removal of pending for specs that require basic metadata.

@samvera/hyrax-code-reviewers

### CODE REVIEW

File: lib/generators/hyrax/collection_resource/templates/collection.rb.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Generated via
+#  `rails generate hyrax:collection_resource <%= class_name %>`
+class <%= class_name %> < Hyrax::PcdmCollection
+  # @note Do not directly update `basic_metadata.yaml`.  It is also used by works.
+  #
+  # To change metadata for collections
+  # * extend by adding fields to `/config/metadata/<%= class_name %>_metadata.yaml`
+  # * remove all or some basic metadata fields
+  #   * comment out or delete the schema include statement for `:basic_metadata`
+  #   * modify the contents of `/config/metadata/<%= class_name %>_metadata.yaml`
```
Comment (cjcolvar) [inline lib/generators/hyrax/collection_resource/templates/collection.rb.erb:12]: Based upon my previous comment.
```suggestion
  # * extend by adding fields to `/config/metadata/<%= file_name %>.yaml`
  # * remove all or some basic metadata fields
  #   * comment out or delete the schema include statement for `:basic_metadata`
  #   * modify the contents of `/config/metadata/<%= file_name %>.yaml`
```

File: spec/test_app_templates/lib/generators/test_app_generator.rb [hunk 2]
```diff
@@ -17,6 +17,33 @@ def add_valkyrie_migrations
...
+      primary: true
+      multiple: true
+  department:
+    type: string
+    form:
+      primary: true
+  course:
+    type: string
+    form:
+      primary: false
```
Comment (cjcolvar) [inline spec/test_app_templates/lib/generators/test_app_generator.rb:40]: Should this also be the config used in dassie?

**Review by cjcolvar — COMMENTED** [review]
This looks really good!  Neat that you're able to turn more tests back on and they pass.
I just had a couple non-functional questions.

File: lib/generators/hyrax/collection_resource/collection_resource_generator.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  def banner
+    if revoking?
+      say_status("info", "DESTROYING VALKYRIE COLLECTION MODEL: #{class_name}", :blue)
+    else
+      say_status("info", "GENERATING VALKYRIE COLLECTION MODEL: #{class_name}", :blue)
+    end
+  end
+
+  def create_metadata_config
+    template('collection_metadata.yaml', File.join('config/metadata/', "#{file_name}.yaml"))
```
[the remark this answers is not in the document]
Comment (elrayle) [reply lib/generators/hyrax/collection_resource/collection_resource_generator.rb:44]: I followed the pattern set by the work resource generator.  For Monograph, it generated `config/metadata/monograph.yaml`.  I think I originally generated `collection_resource_metadata.yaml` and switched it to `collection_resource.yaml` from the precedence set by work resource generator.

File: spec/test_app_templates/lib/generators/test_app_generator.rb (hunk 2 above)
Comment (elrayle) [reply spec/test_app_templates/lib/generators/test_app_generator.rb:40]: I agree, they should match.

Comment (elrayle) [reply spec/test_app_templates/lib/generators/test_app_generator.rb:40]: Updated so both dassie and the test app have these custom attributes.

File: .dassie/app/models/collection_resource.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# Generated via
+#  `rails generate hyrax:collection_resource CollectionResource`
+class CollectionResource < Hyrax::PcdmCollection
+  # @note Do not directly update `basic_metadata.yaml`.  It is also used by works.
+  #
+  # To change metadata for collections
+  # * extend by adding fields to `/config/metadata/CollectionResource_metadata.yaml`
+  # * remove all or some basic metadata fields
+  #   * comment out or delete the schema include statement for `:basic_metadata`
+  #   * modify the contents of `/config/metadata/CollectionResource_metadata.yaml`
```
Comment (cjcolvar) [inline .dassie/app/models/collection_resource.rb:12]: Another small place to make dassie consistent with the generator:
```suggestion
  # * extend by adding fields to `/config/metadata/collection_resource.yaml`
  # * remove all or some basic metadata fields
  #   * comment out or delete the schema include statement for `:basic_metadata`
  #   * modify the contents of `/config/metadata/collection_resource.yaml`
```

**Comment by cjcolvar** [comment]
It looks like the commit you added for #5614 is on this branch too.  Is that intentional?

**Comment by elrayle** [comment]
Hmmm, I thought I had marked #5614 draft, pending the merge of this PR.  I marked it pending now.  My intention is to get this one merged first and then I'll rebase 5614.

**Review by cjcolvar — APPROVED** [review]
Looks good!

**Pull request merged**